### PR TITLE
[ZEPPELIN-5576] Improve Zeppelin CI

### PR DESCRIPTION
**Pull request by zjffdu**
### What is this PR for?

Restructure the CI in `core.yaml`
* Split interpreters CI into 5 groups:
    * jupyter-interpreter, rlang, python
    * non-core interpreters (sh, markdown, etc)
    * spark
    * flink
    * livy
* Run different profiles' spark test in one bulid. Only run it under different python versions.
* Refine the hadoop2 and hadoop3 profiles, use different hadoop profiles in integration test (`SparkIntegrationTest`, `ZeppelinSparkClusterTest`)

### What type of PR is it?
[Improvement ]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-5576

### How should this be tested?
* CI pass except the frontend e2e test (not related)

### Questions:
* Does the licenses files need update? NO
* Is there breaking changes for older versions? NO
* Does this needs documentation? NO

### CODE REVIEW

**Comment by zjffdu** [comment]
@Reamer Do you have time to review it ? It make a lot changes on core CI

File: .github/workflows/core.yml [hunk 1]
```diff
@@ -94,97 +97,25 @@ jobs:
           key: ${{ runner.os }}-zeppelin-${{ hashFiles('**/pom.xml') }}
           restore-keys: |
             ${{ runner.os }}-zeppelin-
+      - name: install environment
+        run: ./mvnw install -DskipTests -DskipRat -am -pl .,zeppelin-interpreter,zeppelin-interpreter-shaded,${INTERPRETERS} -Pscala-2.10 -B
```
Comment (martin-g) [inline .github/workflows/core.yml:101]: Is there a reason to still support/test Scala 2.10 ?
Most of the projects dropped support even for 2.11

File: .github/workflows/frontend.yml [hunk 2]
```diff
@@ -47,6 +47,10 @@ jobs:
         run: ./mvnw -B install -DskipTests -DskipRat -pl ${INTERPRETERS} -Phadoop2 -Pscala-2.11
       - name: Run headless test
         run: xvfb-run --auto-servernum --server-args="-screen 0 1024x768x24" ./mvnw verify -DskipRat -pl zeppelin-web -Phadoop2 -Pscala-2.11 -Pweb-e2e -B
+      - name: Print zeppelin logs
+        if: always()
```
Comment (martin-g) [inline .github/workflows/frontend.yml:51]: Why is this needed ?

File: pom.xml [hunk 3]
```diff
@@ -398,12 +401,27 @@
...
+      <dependency>
+        <groupId>org.apache.hadoop</groupId>
+        <artifactId>${hadoop-client-api.artifact}</artifactId>
+        <version>${hadoop.version}</version>
+        <scope>${hadoop.deps.scope}</scope>
+      </dependency>
+
       <dependency>
         <groupId>org.apache.hadoop</groupId>
         <artifactId>hadoop-client</artifactId>
```
Comment (martin-g) [inline pom.xml:413]: Is this dependency still needed ? It will be a duplicate of the newly added above (line 404-409)

File: pom.xml [hunk 4]
```diff
@@ -532,6 +550,129 @@
...
+          <exclusion>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-compress</artifactId>
+          </exclusion>
+        </exclusions>
+      </dependency>
+
+      <dependency>
+        <groupId>org.apache.hadoop</groupId>
+        <artifactId>hadoop-yarn-client</artifactId>
```
Comment (martin-g) [inline pom.xml:604]: Should this use `${hadoop-client-runtime.artifact}` instead ?

File: pom.xml [hunk 5]
```diff
@@ -997,6 +1215,195 @@
...
+          <exclusion>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+          </exclusion>
+        </exclusions>
+      </dependency>
+
+      <dependency>
+        <groupId>org.apache.hadoop</groupId>
+        <artifactId>${hadoop-client-runtime.artifact}</artifactId>
```
Comment (martin-g) [inline pom.xml:1386]: Maven could complain that this is a duplicate the one above where I asked about `hadoop-client-runtime.artifact`

File: rlang/pom.xml [hunk 6]
```diff
@@ -113,6 +113,13 @@
             <scope>compile</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-common</artifactId>
+            <version>2.7.7</version>
```
Comment (martin-g) [inline rlang/pom.xml:119]: Should this use `${hadoop.version}` ?

File: spark/interpreter/pom.xml [hunk 7]
```diff
@@ -113,6 +113,13 @@
       </exclusions>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-client</artifactId>
+      <version>2.7.7</version>
```
Comment (martin-g) [inline spark/interpreter/pom.xml:119]: `${hadoop.version}` ?

File: spark/interpreter/pom.xml [hunk 8]
```diff
@@ -198,10 +205,23 @@
     <dependency>
       <groupId>org.apache.hadoop</groupId>
       <artifactId>hadoop-client</artifactId>
-      <version>2.6.0</version>
+      <version>2.7.7</version>
```
Comment (martin-g) [inline spark/interpreter/pom.xml:208]: `${hadoop.version}` ?

File: spark/interpreter/pom.xml [hunk 9]
```diff
@@ -198,10 +205,23 @@
...
       <scope>provided</scope>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>2.7.7</version>
```
Comment (martin-g) [inline spark/interpreter/pom.xml:215]: `${hadoop.version}` ?

File: spark/spark-scala-parent/pom.xml [hunk 10]
```diff
@@ -80,6 +80,21 @@
             <scope>provided</scope>
         </dependency>
 
+        <!-- Use provided scope, hadoop dependencies are only for compilation -->
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-client</artifactId>
+            <version>2.7.7</version>
```
Comment (martin-g) [inline spark/spark-scala-parent/pom.xml:87]: `${hadoop.version}` ?

File: zeppelin-interpreter-integration/pom.xml [hunk 11]
```diff
@@ -203,4 +173,105 @@
...
+      <id>hadoop2</id>
+
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+
+      <properties>
+        <hadoop.version>${hadoop2.7.version}</hadoop.version>
+        <curator.version>2.7.1</curator.version>
+        <commons-io.version>2.4</commons-io.version>
```
Comment (martin-g) [inline zeppelin-interpreter-integration/pom.xml:187]: https://search.maven.org/artifact/commons-io/commons-io/2.11.0/jar is the latest one.

File: zeppelin-interpreter-integration/src/test/java/org/apache/zeppelin/integration/ZeppelinSparkClusterTest.java [hunk 12]
```diff
@@ -154,6 +162,11 @@ private void waitForRunning(Paragraph p) {
 
   @Test
   public void scalaOutputTest() throws IOException, InterruptedException {
+    if (!isHadoopVersionMatch()) {
```
Comment (martin-g) [inline zeppelin-interpreter-integration/src/test/java/org/apache/zeppelin/integration/ZeppelinSparkClusterTest.java:165]: IMO this would be better by using JUnit Assumption APIs, e.g. `assumeThat(isHadoopVersionMatch(), "Hadoop version mismatch, skipping the test")`. No need of the `return;`

File: zeppelin-interpreter/pom.xml [hunk 13]
```diff
@@ -233,4 +226,86 @@
...
+    <profile>
+      <id>hadoop2</id>
+
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+      <properties>
+        <hadoop.version>${hadoop2.7.version}</hadoop.version>
+        <curator.version>2.7.1</curator.version>
+        <commons-io.version>2.4</commons-io.version>
```
Comment (martin-g) [inline zeppelin-interpreter/pom.xml:239]: 2.11 ?!

File: zeppelin-server/pom.xml [hunk 14]
```diff
@@ -498,17 +477,95 @@
...
       </properties>
     </profile>
+
+    <profile>
+      <id>hadoop2</id>
+
+      <properties>
+        <hadoop.version>${hadoop2.7.version}</hadoop.version>
+        <curator.version>2.7.1</curator.version>
+        <commons-io.version>2.4</commons-io.version>
```
Comment (martin-g) [inline zeppelin-server/pom.xml:487]: 2.11 ?!

File: zeppelin-zengine/pom.xml [hunk 15]
```diff
@@ -317,4 +312,86 @@
...
+      <id>hadoop2</id>
+
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+
+      <properties>
+        <hadoop.version>${hadoop2.7.version}</hadoop.version>
+        <curator.version>2.7.1</curator.version>
+        <commons-io.version>2.4</commons-io.version>
```
Comment (martin-g) [inline zeppelin-zengine/pom.xml:327]: 2.11 ?!

**Review by martin-g — COMMENTED** [review]
@zjffdu I am not very familiar with Zeppelin but I hope some of my comments are useful!

File: .github/workflows/core.yml (hunk 1 above)
Comment (zjffdu) [reply .github/workflows/core.yml:101]: You are right, but many interpreter modules depends on scala-2.10, and seems there're no maintainers now. We can raise a discussion in community about dropping these un-maintanced interpreter modules.

File: .github/workflows/frontend.yml (hunk 2 above)
Comment (zjffdu) [reply .github/workflows/frontend.yml:51]: Just for diagnose in case test fail, we can check the zeppelin server logs via this action

**Review by Reamer — CHANGES_REQUESTED** [review]
A big PR - I have added a few comments.

File: rlang/pom.xml (hunk 6 above)
Comment (zjffdu) [reply rlang/pom.xml]: Fixed

File: spark/interpreter/pom.xml (hunk 8 above)
Comment (zjffdu) [reply spark/interpreter/pom.xml]: Fixed

File: spark/interpreter/pom.xml (hunk 9 above)
Comment (zjffdu) [reply spark/interpreter/pom.xml]: Fixed

File: spark/spark-scala-parent/pom.xml (hunk 10 above)
Comment (zjffdu) [reply spark/spark-scala-parent/pom.xml]: Fixed

File: zeppelin-interpreter-integration/pom.xml (hunk 11 above)
Comment (zjffdu) [reply zeppelin-interpreter-integration/pom.xml]: Fixed

File: zeppelin-interpreter-integration/src/test/java/org/apache/zeppelin/integration/ZeppelinSparkClusterTest.java (hunk 12 above)
Comment (zjffdu) [reply zeppelin-interpreter-integration/src/test/java/org/apache/zeppelin/integration/ZeppelinSparkClusterTest.java]: Fixed

File: zeppelin-interpreter/pom.xml (hunk 13 above)
Comment (zjffdu) [reply zeppelin-interpreter/pom.xml]: Fixed

File: zeppelin-server/pom.xml (hunk 14 above)
Comment (zjffdu) [reply zeppelin-server/pom.xml]: Fixed

File: zeppelin-zengine/pom.xml (hunk 15 above)
Comment (zjffdu) [reply zeppelin-zengine/pom.xml]: Fixed

File: pom.xml (hunk 3 above)
Comment (zjffdu) [reply pom.xml]: Fixed

File: .github/workflows/core.yml [hunk 16]
```diff
@@ -441,6 +358,11 @@ jobs:
           key: ${{ runner.os }}-zeppelin-${{ hashFiles('**/pom.xml') }}
           restore-keys: |
             ${{ runner.os }}-zeppelin-
+      - name: install environment
+        run: |
+          ./mvnw install -DskipTests -DskipRat -pl livy -am  -B
+          ./testing/downloadSpark.sh "2.2.0" "2.6"
```
[the remark this answers is not in the document]
Comment (zjffdu) [reply .github/workflows/core.yml:364]: We are using an old version of livy (0.5), it may not support newer versions of spark

File: spark/interpreter/pom.xml [hunk 17]
```diff
@@ -113,6 +113,13 @@
       </exclusions>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-client</artifactId>
```
[the remark this answers is not in the document]
Comment (zjffdu) [reply spark/interpreter/pom.xml]: Fixed

File: zeppelin-interpreter-integration/pom.xml [hunk 18]
```diff
@@ -24,7 +24,7 @@
     <artifactId>zeppelin</artifactId>
     <groupId>org.apache.zeppelin</groupId>
     <version>0.11.0-SNAPSHOT</version>
-    <relativePath>..</relativePath>
+    <relativePath>../pom.xml</relativePath>
```
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-interpreter-integration/pom.xml]: Fixed

File: zeppelin-interpreter/pom.xml [hunk 19]
```diff
@@ -233,4 +226,86 @@
...
+
+      <dependencies>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-common</artifactId>
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-yarn-client</artifactId>
```
Comment (Reamer) [inline zeppelin-interpreter/pom.xml:256]: Duplicate declaration.

File: zeppelin-interpreter/pom.xml [hunk 20]
```diff
@@ -233,4 +226,86 @@
...
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-yarn-api</artifactId>
```
Comment (Reamer) [inline zeppelin-interpreter/pom.xml:261]: `hadoop-yarn-api` is a transitive dependency of `hadoop-yarn-client`, it does not need to be defined twice.

File: zeppelin-interpreter/pom.xml [hunk 21]
```diff
@@ -233,4 +226,86 @@
...
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-yarn-common</artifactId>
```
Comment (Reamer) [inline zeppelin-interpreter/pom.xml:266]: `hadoop-yarn-common` is a transitive dependency of `hadoop-yarn-client`, it does not need to be defined twice.

File: zeppelin-interpreter/pom.xml [hunk 22]
```diff
@@ -233,4 +226,86 @@
...
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-yarn-client</artifactId>
```
Comment (Reamer) [inline zeppelin-interpreter/pom.xml:271]: Hadoop libraries are only needed for the YarnUtils class. If I understand YARN correctly, the Hadoop libraries are provided by the system. Therefore, I think we can set the dependency scope to `provided` and a small comment.

**Review by Reamer — CHANGES_REQUESTED** [review]
@zjffdu I have written a partial review. Unfortunately, I don't have time to finish the review this week. Please wait.

File: zeppelin-plugins/notebookrepo/filesystem/pom.xml [hunk 23]
```diff
@@ -43,6 +43,10 @@
             <groupId>org.apache.hadoop</groupId>
             <artifactId>hadoop-client</artifactId>
         </dependency>
+        <dependency>
```
Comment (Reamer) [inline zeppelin-plugins/notebookrepo/filesystem/pom.xml:46]: Based on the Hadoop profile in zeppelin-zengine, the Hadoop lib may or may not be present.
We should also put Hadoop profiles in this `pom.xml` and explicitly set this dependency to `provided`.

File: zeppelin-plugins/launcher/yarn/pom.xml [hunk 24]
```diff
@@ -49,6 +49,11 @@
       <artifactId>hadoop-yarn-api</artifactId>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
```
Comment (Reamer) [inline zeppelin-plugins/launcher/yarn/pom.xml:53]: Based on the Hadoop profile in zeppelin-zengine, the Hadoop lib may or may not be present.
We should also put Hadoop profiles in this `pom.xml` and explicitly set this dependency to `provided`.

File: zeppelin-interpreter-integration/pom.xml [hunk 25]
```diff
@@ -203,4 +172,105 @@
...
+      <id>hadoop2</id>
+
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+
+      <properties>
+        <hadoop.version>${hadoop2.7.version}</hadoop.version>
+        <curator.version>2.7.1</curator.version>
+        <commons-io.version>2.11</commons-io.version>
```
Comment (Reamer) [inline zeppelin-interpreter-integration/pom.xml:186]: I do not see any use of `commons-io.version`, this property should be removed.

File: zeppelin-interpreter-integration/pom.xml [hunk 26]
```diff
@@ -203,4 +172,105 @@
...
+    <profile>
+      <id>hadoop2</id>
+
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+
+      <properties>
+        <hadoop.version>${hadoop2.7.version}</hadoop.version>
+        <curator.version>2.7.1</curator.version>
```
Comment (Reamer) [inline zeppelin-interpreter-integration/pom.xml:185]: I do not see any use of `curator.version`, this property should be removed.

File: zeppelin-interpreter-integration/pom.xml [hunk 27]
```diff
@@ -203,4 +172,105 @@
...
+      </dependencies>
+
+    </profile>
+    <profile>
+      <id>hadoop3</id>
+
+      <properties>
+        <hadoop.version>${hadoop3.2.version}</hadoop.version>
+        <curator.version>2.13.0</curator.version>
+        <kerberos-client.version>2.0.0-M15</kerberos-client.version>
```
Comment (Reamer) [inline zeppelin-interpreter-integration/pom.xml:247]: I do not see any use of `kerberos-client.version`, this property should be removed.

File: zeppelin-interpreter-integration/pom.xml [hunk 28]
```diff
@@ -203,4 +172,105 @@
...
+
+      </dependencies>
+
+    </profile>
+    <profile>
+      <id>hadoop3</id>
+
+      <properties>
+        <hadoop.version>${hadoop3.2.version}</hadoop.version>
+        <curator.version>2.13.0</curator.version>
```
Comment (Reamer) [inline zeppelin-interpreter-integration/pom.xml:246]: I do not see any use of `curator.version`, this property should be removed.

File: zeppelin-interpreter-integration/pom.xml [hunk 29]
```diff
@@ -203,4 +172,105 @@
...
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-runtime.artifact}</artifactId>
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-minicluster.artifact}</artifactId>
```
Comment (Reamer) [inline zeppelin-interpreter-integration/pom.xml:269]: scope `test` should be added

File: zeppelin-interpreter-integration/pom.xml [hunk 30]
```diff
@@ -203,4 +172,105 @@
...
+        <hadoop-client-runtime.artifact>hadoop-client-runtime</hadoop-client-runtime.artifact>
+        <hadoop-client-minicluster.artifact>hadoop-client-minicluster</hadoop-client-minicluster.artifact>
+
+      </properties>
+
+      <dependencies>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-api.artifact}</artifactId>
```
Comment (Reamer) [inline zeppelin-interpreter-integration/pom.xml:259]: I think we can remove this dependency because `hadoop-client-api` is a transitive dependency of `hadoop-client-runtime`.

**Review by Reamer — CHANGES_REQUESTED** [review]
I will do another review after you have made the changes to get a complete overview.
I like the change to move to hadoop shaded jars. If we separate the dependencies of hadoop2 and hadoop3 cleanly in different profiles, it should be very easy to drop hadoop2 in the future to get rid of the dependency jungle.

The new CI structure looks good. :+1:

File: pom.xml (hunk 5 above)
Comment (zjffdu) [reply pom.xml:1386]: They are different, One is `hadoop-yarn-client` another is `hadoop-yarn-api`

File: zeppelin-server/pom.xml [hunk 31]
```diff
@@ -498,17 +477,95 @@
...
+        <hadoop-client-api.artifact>hadoop-client-api</hadoop-client-api.artifact>
+        <hadoop-client-runtime.artifact>hadoop-client-runtime</hadoop-client-runtime.artifact>
+        <hadoop-client-minicluster.artifact>hadoop-client-minicluster</hadoop-client-minicluster.artifact>
+
+      </properties>
+
       <dependencies>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
```
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-server/pom.xml:546]: It is only true when it is hadoop3, but not true for hadoop2

File: zeppelin-interpreter-integration/pom.xml (hunk 25 above)
Comment (zjffdu) [reply zeppelin-interpreter-integration/pom.xml]: fixed

File: zeppelin-interpreter-integration/pom.xml (hunk 26 above)
Comment (zjffdu) [reply zeppelin-interpreter-integration/pom.xml]: fixed

File: zeppelin-interpreter-integration/pom.xml (hunk 27 above)
Comment (zjffdu) [reply zeppelin-interpreter-integration/pom.xml:247]: fixed

File: zeppelin-interpreter-integration/pom.xml (hunk 28 above)
Comment (zjffdu) [reply zeppelin-interpreter-integration/pom.xml:246]: fixed

File: zeppelin-interpreter/pom.xml (hunk 20 above)
Comment (zjffdu) [reply zeppelin-interpreter/pom.xml]: fixed

File: zeppelin-interpreter/pom.xml (hunk 21 above)
Comment (zjffdu) [reply zeppelin-interpreter/pom.xml]: fixed

File: zeppelin-server/pom.xml [hunk 32]
```diff
@@ -498,17 +477,95 @@
...
+        </dependency>
+      </dependencies>
+
+    </profile>
     <profile>
       <id>hadoop3</id>
+
+      <properties>
+        <hadoop.version>${hadoop3.2.version}</hadoop.version>
+        <curator.version>2.13.0</curator.version>
```
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-server/pom.xml]: fixed

File: zeppelin-server/pom.xml [hunk 33]
```diff
@@ -498,17 +477,95 @@
...
+        <kerberos-client.version>2.0.0-M15</kerberos-client.version>
```
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-server/pom.xml]: Fixed

File: zeppelin-interpreter/pom.xml [hunk 34]
```diff
@@ -233,4 +226,86 @@
...
+      <dependencies>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-api.artifact}</artifactId>
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-runtime.artifact}</artifactId>
```
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-interpreter/pom.xml:276]: The reason is that allow the user to control whether package hadoop in zeppelin distribution via the property `hadoop.deps.scope` defined in root pom.xml

**Comment by zjffdu** [comment]
@Reamer Any more comments?

**Comment by Reamer** [comment]
> @Reamer Any more comments?

Yes let me review this agian.

File: zeppelin-interpreter/pom.xml [hunk 35]
```diff
@@ -233,4 +226,86 @@
...
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-minicluster.artifact}</artifactId>
```
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-interpreter/pom.xml:281]: Still valid.

File: zeppelin-server/pom.xml [hunk 36]
```diff
@@ -293,30 +293,9 @@
...
-      <exclusions>
-        <!-- Provided by javax-websocket-server-impl -->
-        <exclusion>
-          <groupId>org.ow2.asm</groupId>
-          <artifactId>asm</artifactId>
-        </exclusion>
-      </exclusions>
+      <groupId>org.apache.directory.server</groupId>
+      <artifactId>apacheds-kerberos-codec</artifactId>
+      <version>2.0.0-M15</version>
```
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-server/pom.xml:298]: Still valid.

File: zeppelin-server/pom.xml (hunk 31 above)
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-server/pom.xml:508]: Your right, but we are in the profile `hadoop3`, therefore is doesn't effect `hadoop2`

File: zeppelin-zengine/pom.xml [hunk 37]
```diff
@@ -317,4 +312,86 @@
...
+      </properties>
+      <dependencies>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-common</artifactId>
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-yarn-client</artifactId>
```
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-zengine/pom.xml:338]: still valid

File: zeppelin-zengine/pom.xml [hunk 38]
```diff
@@ -317,4 +312,86 @@
...
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-yarn-api</artifactId>
```
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-zengine/pom.xml:343]: still valid

File: zeppelin-zengine/pom.xml [hunk 39]
```diff
@@ -317,4 +312,86 @@
...
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-yarn-common</artifactId>
```
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-zengine/pom.xml:348]: still valid

File: zeppelin-zengine/pom.xml [hunk 40]
```diff
@@ -317,4 +312,86 @@
...
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-runtime.artifact}</artifactId>
+        </dependency>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-minicluster.artifact}</artifactId>
```
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-zengine/pom.xml:383]: still valid

File: zeppelin-zengine/pom.xml [hunk 41]
```diff
@@ -317,4 +312,86 @@
...
+        <hadoop-client-runtime.artifact>hadoop-client-runtime</hadoop-client-runtime.artifact>
+        <hadoop-client-minicluster.artifact>hadoop-client-minicluster</hadoop-client-minicluster.artifact>
+
+      </properties>
+
+      <dependencies>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-api.artifact}</artifactId>
```
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-zengine/pom.xml:373]: still valid

File: zeppelin-plugins/notebookrepo/filesystem/pom.xml (hunk 23 above)
Comment (Reamer) [reply zeppelin-plugins/notebookrepo/filesystem/pom.xml:46]: still valid

File: zeppelin-plugins/launcher/yarn/pom.xml (hunk 24 above)
Comment (Reamer) [reply zeppelin-plugins/launcher/yarn/pom.xml:53]: still valid

File: zeppelin-interpreter/pom.xml (hunk 34 above)
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-interpreter/pom.xml:276]: I understand that. But doesn't YARN provide the Hadoop libraries? The only use of Hadoop libraries are in the `YarnUtils` class. So I think that Zeppelin interpreter should not provide Hadoop libraries.

File: zeppelin-interpreter/pom.xml [hunk 42]
```diff
@@ -233,4 +226,86 @@
...
+        <hadoop-client-runtime.artifact>hadoop-client-runtime</hadoop-client-runtime.artifact>
+        <hadoop-client-minicluster.artifact>hadoop-client-minicluster</hadoop-client-minicluster.artifact>
+
+      </properties>
+
+      <dependencies>
+
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>${hadoop-client-api.artifact}</artifactId>
```
[the remark this answers is not in the document]
Comment (Reamer) [reply zeppelin-interpreter/pom.xml:271]: still valid

**Review by Reamer — CHANGES_REQUESTED** [review]
@zjffdu 
I think most of my comments are still valid.

File: zeppelin-interpreter-integration/pom.xml (hunk 30 above)
Comment (zjffdu) [reply zeppelin-interpreter-integration/pom.xml]: Fixed

File: zeppelin-interpreter/pom.xml (hunk 34 above)
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-interpreter/pom.xml:276]: As I mentioned in the above comment, user can choose whether packaging hadoop into Zeppelin distribution. This is just like what spark did (spark has 2 distribution, one with hadoop libraries, another without hadoop libraries)

File: zeppelin-server/pom.xml (hunk 36 above)
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-server/pom.xml:298]: Fixed

File: zeppelin-server/pom.xml (hunk 31 above)
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-server/pom.xml:508]: Fixed

File: zeppelin-zengine/pom.xml (hunk 37 above)
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-zengine/pom.xml:322]: Fixed

File: zeppelin-zengine/pom.xml (hunk 41 above)
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-zengine/pom.xml:357]: Fixed

File: zeppelin-zengine/pom.xml (hunk 40 above)
[the remark this answers is not in the document]
Comment (zjffdu) [reply zeppelin-zengine/pom.xml:367]: Fixed

**Comment by zjffdu** [comment]
@Reamer  Sorry for late response, I think all the comments are addressed now, could you help do another round of review?

**Comment by Reamer** [comment]
I will have a look at it next week.

**Review by Reamer — APPROVED** [review]
A very complex PR. I hope that I have found a lot. For me, this PR looks good. The new CI jobs are up and running, thanks for that.
